### PR TITLE
Random images

### LINKED ISSUE #26: Configurable random images on home page
Need to provide a database model for home page images (via depot) & descriptions

Select randomly when the page is loaded

### PR DESCRIPTION
**Pull request by edwargix**
Resolves #26

### CODE REVIEW

File: acmwebsite/controllers/root.py [hunk 1]
```diff
@@ -59,7 +59,23 @@ def index(self):
         meetings = DBSession.query(Meeting).filter(
             Meeting.date > datetime.datetime.now() - datetime.timedelta(hours=3)
         ).order_by(Meeting.date).limit(2)
-        return dict(page='index', meetings=meetings)
+        try:
```
Comment (sumnerevans) [inline acmwebsite/controllers/root.py]: I don't like this. I think we should be doing an `if` on the count of the Banner query.

File: acmwebsite/controllers/root.py [hunk 2]
```diff
@@ -59,7 +59,23 @@ def index(self):
...
+            banner = random.choice(DBSession.query(Banner).all())
+        except:
+            banner = type(
```
Comment (sumnerevans) [inline acmwebsite/controllers/root.py]: Hmm... seems like there must be a better way of doing this... @jackrosenthal

File: acmwebsite/controllers/root.py [hunk 3]
```diff
@@ -59,7 +59,23 @@ def index(self):
...
+                '',
+                (object,),
+                dict(
+                    description='Image above: ACM Members and other clubs at a'
+                                'Mozilla Campus Clubs training event',
+                    photo=type(
+                        'obj',
+                        (object,),
+                        {'url': '/img/banner.jpg'}
```
Comment (sumnerevans) [inline acmwebsite/controllers/root.py]: This should be called `default_banner.jpg`.

Comment (jackrosenthal) [reply acmwebsite/controllers/root.py]: Default should be nothing... not the image that just so happens to be there.

Handle default logic (showing nothing when banner is `None`) in template...

File: acmwebsite/model/banner.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+from sqlalchemy import Column
+from sqlalchemy.types import Integer, Unicode
+from depot.fields.sqlalchemy import UploadedFileField
+from depot.fields.specialized.image import UploadedImageWithThumb
+
+
+class Banner(DeclarativeBase):
+    __tablename__ = 'banner'
+
+    uid = Column(Integer, autoincrement=True, primary_key=True)
```
Comment (jackrosenthal) [inline acmwebsite/model/banner.py]: Why `uid` than `id`?

File: acmwebsite/templates/index.xhtml [hunk 5]
```diff
@@ -10,10 +10,8 @@
     <div class="row">
       <div class="col-md-7">
         <div class="home-banner">
-          <img class="home-banner-img" src="${tg.url('/img/banner.jpg')}" />
-          <small>
-          Image above: ACM Members and other clubs at a Mozilla Campus Clubs training event
-          </small>
+          <img class="home-banner-img" src="${banner.photo.url}" />
+          <small py:content="banner.description"></small>
```
[the remark this answers is not in the document]
Comment (jackrosenthal) [reply acmwebsite/templates/index.xhtml]: Markdown that description?

File: acmwebsite/controllers/root.py [hunk 6]
```diff
@@ -23,7 +23,7 @@
 from acmwebsite.controllers.schedule import ScheduleController
 from acmwebsite.controllers.survey import SurveysController
 
-import datetime
+import datetime, random
```
Comment (jackrosenthal) [inline acmwebsite/controllers/root.py]: Let the SQL server handle the random image selection for you :)

File: .gitignore [hunk 7]
```diff
@@ -19,3 +20,6 @@ pip-log.txt
 # paste config files
 development.ini
 production.ini
+
+# Rope
+.ropeproject
```
Comment (jackrosenthal) [inline .gitignore]: What's this needed for?

File: .gitignore [hunk 8]
```diff
@@ -3,6 +3,7 @@
 
 # Default development database
 devdata.db
+devdata.db-journal
```
Comment (jackrosenthal) [inline .gitignore]: Not sure when this file gets created... have not seen on my system? Are `-journal` files specific to your system (if so, add to your own global `.gitignore`, not ours)

Comment (sumnerevans) [reply .gitignore]: These are created when SQLite is doing operations. Sometimes it doesn't clean up properly, so this is necessary.

File: .gitignore (hunk 7 above)
Comment (sumnerevans) [reply .gitignore]: Me.

Comment (jackrosenthal) [reply .gitignore]: Yuck! _Please just add to your own global gitignore!!!_

No reason to clutter up our repository with harmful crap like this!!!

Comment (sumnerevans) [reply .gitignore]: Done.

**Comment by edwargix** [comment]
I think I correctly changed what you all wanted. Let me know if this isn't the case.

File: acmwebsite/templates/index.xhtml [hunk 9]
```diff
@@ -9,11 +9,9 @@
   <body py:block="body" py:strip="True">
     <div class="row">
       <div class="col-md-7">
-        <div class="home-banner">
-          <img class="home-banner-img" src="${tg.url('/img/banner.jpg')}" />
-          <small>
-          Image above: ACM Members and other clubs at a Mozilla Campus Clubs training event
-          </small>
+        <div py:if="banner is not None" class="home-banner">
```
Comment (sumnerevans) [inline acmwebsite/templates/index.xhtml]: I don't really like the look of this. There is a lot of whitespace to the left of the "Mines ACM Student Chapter" text. @jackrosenthal, what do you think?

If we want the "Mines ACM Student Chapter" text left-aligned when there is no banner image, this `py:if` statement should be on the `div` above this (with class `col-md-7`) so that the "Mines ACM Student Chapter" text is the first `div` inside of this row.

File: acmwebsite/model/banner.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+from depot.fields.sqlalchemy import UploadedFileField
+from depot.fields.specialized.image import UploadedImageWithThumb
+
+
+class Banner(DeclarativeBase):
+    __tablename__ = 'banner'
+
+    id = Column(Integer, autoincrement=True, primary_key=True)
+    photo = Column(UploadedFileField(upload_type=UploadedImageWithThumb))
+    description = Column(Unicode(255), unique=True)
```
Comment (sumnerevans) [inline acmwebsite/model/banner.py]: @jackrosenthal, what DB is backing this? Postgresql?

If so, @edwargix, use `Text` instead of `Unicode(255)` for this field, we don't want to limit descriptions to 255 chars.

File: acmwebsite/templates/index.xhtml [hunk 11]
```diff
@@ -9,11 +9,9 @@
...
+          <img class="home-banner-img" src="${banner.photo.url}" />
+          <small py:if="banner.description" py:content="banner.description"></small>
```
Comment (sumnerevans) [inline acmwebsite/templates/index.xhtml]: As @jackrosenthal, mentioned in an earlier review, this should be markdownified. (See `templates/meeting_list.xhtml:5` for an example.

File: acmwebsite/model/banner.py (hunk 10 above)
Comment (jackrosenthal) [reply acmwebsite/model/banner.py]: Yes, although it is good to keep a limit on the size (important for some other databases, MySQL cough cough, and I don't want to break compatibility for this one thing).

255 bytes is a bit small tho

File: acmwebsite/templates/index.xhtml (hunk 9 above)
Comment (jackrosenthal) [reply acmwebsite/templates/index.xhtml]: Honestly no issues with it ... it is intended that a picture show there, and if there's not, it's nice keeping it consistent with the width of the text (intention is that you would not need to make banners during development).

Maybe toss a bootstrap "jumbotron" in the oppisite column when there is no image? Not very important.

File: setup.py [hunk 12]
```diff
@@ -50,6 +50,7 @@
     "tw2.forms",
     "WebHelpers2",
     "zope.sqlalchemy >= 0.4",
+    "Pillow",
```
Comment (jackrosenthal) [inline setup.py]: where is this used?

**Comment by jackrosenthal** [comment]
Also, we can get rid of `banner.jpg` now.

File: acmwebsite/model/banner.py (hunk 10 above)
Comment (sumnerevans) [reply acmwebsite/model/banner.py]: @jackrosenthal , proposals for a more appropriate size? I'm thinking 512-1000 is fine.

File: setup.py (hunk 12 above)
Comment (edwargix) [reply setup.py]: `Pilloe` is needed to upload images. See this LOC: https://github.com/ColoradoSchoolOfMines/acm-website/blob/8729e9f163e2ac883949cb201e215c5d2c68e646/acmwebsite/model/banner.py#L12

File: acmwebsite/templates/index.xhtml [hunk 13]
```diff
@@ -11,7 +11,7 @@
       <div class="col-md-7">
         <div py:if="banner is not None" class="home-banner">
           <img class="home-banner-img" src="${banner.photo.url}" />
-          <small py:if="banner.description" py:content="banner.description"></small>
+          <small py:if="banner.description">${h.markdown(banner.description)}</small>
```
Comment (sumnerevans) [inline acmwebsite/templates/index.xhtml]: You can still use `py:content` here:

    <small py:if="banner.description" py:content="h.markdown(banner.description)"</small>

File: acmwebsite/model/banner.py [hunk 14]
```diff
@@ -10,4 +10,4 @@ class Banner(DeclarativeBase):
 
     id = Column(Integer, autoincrement=True, primary_key=True)
     photo = Column(UploadedFileField(upload_type=UploadedImageWithThumb))
-    description = Column(Unicode(255), unique=True)
+    description = Column(Text, unique=True)
```
Comment (sumnerevans) [inline acmwebsite/model/banner.py]: Per https://github.com/ColoradoSchoolOfMines/acm-website/pull/32#discussion_r154513923, I think we want to still use `Unicode`, but just with like a 2048 char limit.

Deferring to @jackrosenthal on this.

File: acmwebsite/templates/index.xhtml (hunk 9 above)
Comment (edwargix) [reply acmwebsite/templates/index.xhtml]: How does this look? (I just changed the class of the div on the right from col-md-5 to jumbotron, as @jackrosenthal suggested)

No banner: acm-no-banner

With banner: acm-with-banner

Comment (sumnerevans) [reply acmwebsite/templates/index.xhtml]: @edwargix , I like the look of that overall. I think you need to nest the jumbotron inside of a `col-md-12` so that it doesn't spill over the sides of the page (note that it's wider than the nav bar and the two boxes below).

**Comment by edwargix** [comment]
Once we perfect this, do we want a rebase of the branch to fixup the commits and have the history a little cleaner? (i.e. not have changes that get reverted 2 commits later)

**Comment by sumnerevans** [comment]
@edwargix: Yes. Rebase for sure. Clean up as much as necessary, but don't go overboard on it. (It's fine to have a couple commits see #33 for example.)

**Comment by jackrosenthal** [comment]
Please be careful about this rebasing Sumner is reccomending. Generally, it's _not a good thing_ if you need to force push to the repo (it's like trying to rewrite history!) Especially, never force push to `master` without contacting me first.

Sumner has been known to over-rebase in the past. Just because he does it a lot does not mean it's a good thing...

**Comment by edwargix** [comment]
Would it be best to rebase onto a new branch, like `random-images-rebase`?

**Comment by sumnerevans** [comment]
The master branch is already protected and no force pushes are allowed. Merges are fine. Honestly, I actually prefer it, I thought that you liked it better (maybe I'm confusing you and Sam)

**Comment by jackrosenthal** [comment]
If you cannot justify _why_ you should rebase, then don't. It's pretty simple.

__What will you get out of rebasing?__ It makes sense for things like cleaning up binaries accidentally comitted, but you definately overuse it. Essentially, just justify _why_ you want to rewrite history, then it should be OK. But not allowed without justification.

**Comment by edwargix** [comment]
Ok the rebase is done and can be merged unless there's something else I'm missing. I have a local copy of the branch prior to the rebase, so if anyone needs that just let me know.

**Comment by jackrosenthal** [comment]
PLEASE DO NOT REBASE WITHOUT JUSTIFYING FIRST!!!

**Comment by edwargix** [comment]
I rebased because there was a lot of back-and-forth between commits. The end code is identical and works the same.

**Comment by jackrosenthal** [comment]
Yes, but from what I could tell, the commits told a history. You only need to rebase if binaries are committed, you're smashing up wip commits, or other things like that. There's no reason to destroy incremental steps of history in this thing otherwise.

**Comment by sumnerevans** [comment]
I'll take the blame on this one. But Jack, your hard-line on no rebasing is a detriment to the cleanliness of the commit tree.

**Comment by edwargix** [comment]
If we prefer the history with wip commits, I can reset the head of this branch back to that.

**Comment by jackrosenthal** [comment]
The commit tree tells history. If that is destroyed, then the commit tree is **uncleanly**, the history has been removed. There are no WIP commits from what I can tell, so please do not rebase.

**Comment by edwargix** [comment]
I already did rebase, so should I reset this branch's head back to what it was before?

**Comment by jackrosenthal** [comment]
Yes

**Comment by edwargix** [comment]
Done

**Comment by sumnerevans** [comment]
@jackrosenthal, it's perfectly fine for you to take a hard line on these matters of project management. However, without a `CONTRIBUTING.md` specifying your direction, I cannot guarantee that I or anyone else will do exactly what you want.

Again, I'm taking the blame on this one, but you continue to take sides on issues and expect us to be able to read your mind which is not a sustainable project management strategy.

(Sorry for having to be caught up in this @edwargix.)

**Comment by jackrosenthal** [comment]
Sorry, I had no idea that one would do anything without justification, let alone rebase without justification, when working on the repository. Doing things without justification is a pretty good definition of bigotry.

**Comment by sumnerevans** [comment]
@jackrosenthal, it is equally, if not more, bigoted to consider a justification you don't agree with to be invalid and nonexistent. @edwargix and I made it clear that the reason to rebase was to remove back and forth commits where things were changed and then immediately reverted back.

**Comment by jackrosenthal** [comment]
Back and forth commits show history. They weren't WIP commits that were quickly changed. They weren't binary commits accidentally commited. It was _good_ these were separate commits!

**Comment by samsartor** [comment]
@jackrosenthal I have been staying out of this so far, but I have to agree with @sumnerevans on this one. Different repos use different strategies. I have worked on projects where rebasing and squashing was required before merging. If you have a preference you absolutely *must* put it in `CONTRIBUTING.md`.

**Comment by jackrosenthal** [comment]
Never assume.

Repositories which require that specify that. Repositories which don't require that (and don't want it) should not be required to specify that simply because other projects implement that.

In other words, the lack of a specification means something too!

**Comment by sumnerevans** [comment]
@jackrosenthal, 

Re "back and forth": case and point. The word "good" always implies judgment. That is your *opinion*. This is your project, so you have every right to impose your opinion on this project. However, assuming that we are all on the same page is extremely narrow-minded.

Re "never assume": you should never assume that we can read your mind. If rebasing is something you are so passionate about, then why would you *not* want to specify it it a `CONTRIBUTING.md`? If you want to take a hard line on something, specify it in writing. PEP8 is an example of a document which clearly defines where hard lines exist, even if they are obvious (i.e. no tabs, use 4 spaces for indentation).

**Comment by samsartor** [comment]
@sumnerevans @jackrosenthal PR comments are not the right place to have this argument. This is what chats are for.

**Comment by sumnerevans** [comment]
On the actual PR, I'm not sure the best way to fix this, but there are some problems on small screens with this style. (Note that the picture is not inside the Jumbotron. Not sure if we want that or not.)



Also, the text seems a bit big, and there is possibly too small of a margin with the picture.



I think we should go with what Jack suggested in this thread: https://github.com/ColoradoSchoolOfMines/acm-website/pull/32#discussion_r154493443, and revert back to what we had before (just whitespace there when no image is found).

What's your say on this @jackrosenthal?

**Comment by jackrosenthal** [comment]
> What's your say on this @jackrosenthal?

2017-12-15-192745_1466x736_scrot

Pay careful attention to "when there is no image".

**Comment by jackrosenthal** [comment]
Honestly think we should just revert back to what we had before (a very _good_ reason to keep commit history on things like this).

If there was to be a jumbotron, it would have been where the missing picture is, not where the text on the right is.

File: acmwebsite/model/banner.py [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+from depot.fields.sqlalchemy import UploadedFileField
+from depot.fields.specialized.image import UploadedImageWithThumb
+
+
+class Banner(DeclarativeBase):
+    __tablename__ = 'banner'
+
+    id = Column(Integer, autoincrement=True, primary_key=True)
+    photo = Column(UploadedFileField(upload_type=UploadedImageWithThumb))
+    description = Column(Unicode(2048), unique=True)
```
Comment (sumnerevans) [inline acmwebsite/model/banner.py]: Just realized, this is going to require a database upgrade script.

Comment (sumnerevans) [reply acmwebsite/model/banner.py]: Actually, this entire table needs an upgrade script.

File: acmwebsite/templates/index.xhtml [hunk 16]
```diff
@@ -9,11 +9,9 @@
...
     <div class="row">
       <div class="col-md-7">
-        <div class="home-banner">
-          <img class="home-banner-img" src="${tg.url('/img/banner.jpg')}" />
-          <small>
-          Image above: ACM Members and other clubs at a Mozilla Campus Clubs training event
-          </small>
+        <div py:if="banner is not None" class="home-banner">
+          <img class="home-banner-img" src="${banner.photo.url}" />
+          <small py:if="banner.description" py:content="h.markdown(banner.description)"></small>
```
Comment (sumnerevans) [inline acmwebsite/templates/index.xhtml]: Just an FYI (I don't see any reason to change this), when you don't have any content between the tags like this, you can make it a single tag:

    <small py::if="..." py:content="..." />

**Comment by jackrosenthal** [comment]
Looks good to me

File: migration/versions/bbd3bba567c1_banners.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+import sqlalchemy as sa
+from depot.fields.sqlalchemy import UploadedFileField
+from depot.fields.specialized.image import UploadedImageWithThumb
+
+
+def upgrade():
+    op.create_table(
+        'banner',
+        sa.Column('id', sa.Integer(), autoincrement=True, primary_key=True),
+        sa.Column('photo',
```
Comment (sumnerevans) [inline migration/versions/bbd3bba567c1_banners.py]: There seems to be a problem with this. When I edit just the `description` from the admin page, it deletes the photo.

Comment (jackrosenthal) [reply migration/versions/bbd3bba567c1_banners.py]: This is a side effect of the TurboGears admin interface when used with depot...

Comment (sumnerevans) [reply migration/versions/bbd3bba567c1_banners.py]: @jackrosenthal , hmmm... seems like a pretty bad situation. What if the person editing no longer has access to the original file? Is there literally no way around this? If not, then I wonder if we need to consider a better, more stable solution?

Comment (jackrosenthal) [reply migration/versions/bbd3bba567c1_banners.py]: Issue is with the way TG admin interface works. Essentially, I think you end up overwriting all columns when updating an entry (there is no interface to just edit certain columns). If you want this fixed, you could always consider forking the TG admin interface and submit a PR, but I don't consider it a big enough deal.

tl;dr, not this project. Submit the issue elsewhere.

Comment (sumnerevans) [reply migration/versions/bbd3bba567c1_banners.py]: Fair enough.

**Comment by jackrosenthal** [comment]
Ready to merge?
